ee/hogai/graph/base/node.py

```diff
@@ -97,23 +74,78 @@ def context_manager(self) -> AssistantContextManager:
     @property
     def node_name(self) -> str:
         config_name: str | None = None
-        if self._config:
-            config_name = self._config["metadata"].get("langgraph_node")
+        if self._config and (metadata := self._config.get("metadata")):
+            config_name = metadata.get("langgraph_node")
             if config_name is not None:
                 config_name = str(config_name)
         return config_name or self.__class__.__name__
 
+    @property
+    def node_path(self) -> tuple[NodePath, ...]:
+        return (*self._node_path, NodePath(name=self.node_name))
+
+    async def _execute(self, state: StateType, config: RunnableConfig) -> PartialStateType | None:
+        try:
+            return await self._arun_with_context(state, config)
+        except NotImplementedError:
+            pass
+        return await database_sync_to_async(self._run_with_context, thread_sensitive=False)(state, config)
+
+    def _run_with_context(self, state: StateType, config: RunnableConfig) -> PartialStateType | None:
+        with set_node_path(self.node_path):
+            return self.run(state, config)
+
+    async def _arun_with_context(self, state: StateType, config: RunnableConfig) -> PartialStateType | None:
+        with set_node_path(self.node_path):
+            return await self.arun(state, config)
+
+
+class BaseAssistantNode(BaseExecutableAssistantNode[StateType, PartialStateType]):
+    """Assistant node with dispatching and conversation cancellation support."""
+
+    _is_context_path_used: bool = False
+    """Whether the constructor set the node path or the context path is used"""
```

Comment:
I'd need more information about what a node path is vs what a context path is, I understand it means that the path is provided, or it's constructed locally, but the comment is unclear